hw-model/src/lib.rs

```diff
@@ -871,162 +706,92 @@ pub trait HwModel {
         &mut self,
-        mut req: R,
-    ) -> std::result::Result<R::Resp, ModelError> {
-        if mem::size_of::<R>() < mem::size_of::<MailboxReqHeader>() {
-            return Err(ModelError::MailboxReqTypeTooSmall);
-        }
-        if mem::size_of::<R::Resp>() < mem::size_of::<MailboxRespHeader>() {
-            return Err(ModelError::MailboxRespTypeTooSmall);
-        }
-        if R::Resp::MIN_SIZE < mem::size_of::<MailboxRespHeader>() {
-            return Err(ModelError::MailboxRespTypeTooSmall);
-        }
-        let (header_bytes, payload_bytes) = req
-            .as_bytes_mut()
-            .split_at_mut(mem::size_of::<MailboxReqHeader>());
-
-        let mut header = MailboxReqHeader::read_from(header_bytes as &[u8]).unwrap();
-        header.chksum = api::calc_checksum(R::ID.into(), payload_bytes);
-        header_bytes.copy_from_slice(header.as_bytes());
-
-        let Some(response_bytes) = self.mailbox_execute(R::ID.into(), req.as_bytes())? else {
-            return Err(ModelError::MailboxNoResponseData);
-        };
-        if response_bytes.len() < R::Resp::MIN_SIZE
-            || response_bytes.len() > mem::size_of::<R::Resp>()
-        {
-            return Err(ModelError::MailboxUnexpectedResponseLen {
-                expected_min: R::Resp::MIN_SIZE as u32,
-                expected_max: mem::size_of::<R::Resp>() as u32,
-                actual: response_bytes.len() as u32,
-            });
-        }
-
-        let mut response = R::Resp::new_zeroed();
-        response.as_bytes_mut()[..response_bytes.len()].copy_from_slice(&response_bytes);
-
-        let response_header =
-            MailboxRespHeader::read_from_prefix(response_bytes.as_slice()).unwrap();
-        let actual_checksum = calc_checksum(0, &response_bytes[4..]);
-        if actual_checksum != response_header.chksum {
-            return Err(ModelError::MailboxRespInvalidChecksum {
-                expected: response_header.chksum,
-                actual: actual_checksum,
-            });
-        }
-        if response_header.fips_status != MailboxRespHeader::FIPS_STATUS_APPROVED {
-            return Err(ModelError::MailboxRespInvalidFipsStatus(
-                response_header.fips_status,
-            ));
-        }
-        Ok(response)
+        req: R,
+        response: &'a mut R::Resp,
+    ) -> std::result::Result<&'a R::Resp, ModelError> {
+        SocManager::mailbox_exec_req(self, req, response).map_err(ModelError::from)
     }
 
     /// Executes `cmd` with request data `buf`. Returns `Ok(Some(_))` if
     /// the uC responded with data, `Ok(None)` if the uC indicated success
     /// without data, Err(ModelError::MailboxCmdFailed) if the microcontroller
     /// responded with an error, or other model errors if there was a problem
     /// communicating with the mailbox.
-    fn mailbox_execute(
+    fn mailbox_execute<'r>(
+        &mut self,
+        cmd: u32,
+        buf: &[u8],
+        resp_data: &'r mut MboxBuffer,
```

Comment:
Should probably just prefer `&mut [u8]` here instead of a fixed-size type in the API.

Reply:
+1

Reply:
Done.